util/gitutil/git_url_test.go

```diff
@@ -1,106 +1,146 @@
 		},
 		{
-			url:      "ssh://git@github.com:22/moby/buildkit.git",
-			protocol: SSHProtocol,
-			host:     "github.com:22",
-			path:     "/moby/buildkit.git",
-			user:     "git",
+			url: "ssh://git@github.com/moby/buildkit.git",
+			result: GitURL{
+				Scheme: SSHProtocol,
+				Host:   "github.com",
+				Path:   "/moby/buildkit.git",
+				User:   url.User("git"),
+			},
 		},
 		{
-			url:      "git@github.com:moby/buildkit.git",
-			protocol: SSHProtocol,
-			host:     "github.com",
-			path:     "/moby/buildkit.git",
-			user:     "git",
+			url: "ssh://git@github.com:22/moby/buildkit.git",
+			result: GitURL{
+				Scheme: SSHProtocol,
+				Host:   "github.com:22",
+				Path:   "/moby/buildkit.git",
+				User:   url.User("git"),
+			},
 		},
 		{
-			url:      "git@github.com:moby/buildkit.git#v1.0.0",
-			protocol: SSHProtocol,
-			host:     "github.com",
-			path:     "/moby/buildkit.git",
-			fragment: "v1.0.0",
-			user:     "git",
+			url: "git@github.com:moby/buildkit.git",
+			result: GitURL{
+				Scheme: SSHProtocol,
+				Host:   "github.com",
+				Path:   "moby/buildkit.git",
+				User:   url.User("git"),
+			},
 		},
 		{
-			url:      "nonstandarduser@example.com:/srv/repos/weird/project.git",
-			protocol: SSHProtocol,
-			host:     "example.com",
-			path:     "/srv/repos/weird/project.git",
-			user:     "nonstandarduser",
+			url: "git@github.com:moby/buildkit.git#v1.0.0",
+			result: GitURL{
+				Scheme:   SSHProtocol,
+				Host:     "github.com",
+				Path:     "moby/buildkit.git",
+				Fragment: &GitURLFragment{Ref: "v1.0.0"},
+				User:     url.User("git"),
+			},
 		},
 		{
-			url:      "ssh://root@subdomain.example.hostname:2222/root/my/really/weird/path/foo.git",
-			protocol: SSHProtocol,
-			host:     "subdomain.example.hostname:2222",
-			path:     "/root/my/really/weird/path/foo.git",
-			user:     "root",
+			url: "git@github.com:moby/buildkit.git#v1.0.0:hack",
+			result: GitURL{
+				Scheme:   SSHProtocol,
+				Host:     "github.com",
+				Path:     "moby/buildkit.git",
+				Fragment: &GitURLFragment{Ref: "v1.0.0", Subdir: "hack"},
+				User:     url.User("git"),
+			},
 		},
 		{
-			url:      "git://host.xz:1234/path/to/repo.git",
-			protocol: GitProtocol,
-			host:     "host.xz:1234",
-			path:     "/path/to/repo.git",
+			url: "nonstandarduser@example.com:/srv/repos/weird/project.git",
+			result: GitURL{
+				Scheme: SSHProtocol,
+				Host:   "example.com",
+				Path:   "/srv/repos/weird/project.git",
+				User:   url.User("nonstandarduser"),
+			},
+		},
+		{
+			url: "ssh://root@subdomain.example.hostname:2222/root/my/really/weird/path/foo.git",
+			result: GitURL{
+				Scheme: SSHProtocol,
+				Host:   "subdomain.example.hostname:2222",
+				Path:   "/root/my/really/weird/path/foo.git",
+				User:   url.User("root"),
+			},
+		},
+		{
+			url: "git://host.xz:1234/path/to/repo.git",
+			result: GitURL{
+				Scheme: GitProtocol,
+				Host:   "host.xz:1234",
+				Path:   "/path/to/repo.git",
+			},
```

Comment:
should we add another test for non-standard port, but home directory?

e.g.

```
			url: "ssh://someuser@192.168.0.123:456/~/repo-in-my-home-dir.git",
			result: GitURL{
				Scheme: SSHProtocol,
				Host:   "192.168.0.123:456",
				Path:   "/~/repo-in-my-home-dir.git",
                                User: url.User("someuser"),
			},
```

Reply:
Good call, thanks for the example, I've added it in verbatim :tada: